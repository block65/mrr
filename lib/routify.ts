import type { RouteParams, ExtractRouteParams } from './types.js';
import { urlObjectAssign } from './util.js';

type QueryParams = Record<string, string>;

export interface Routified<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => URL;
}

function interpolate<T extends RouteParams>(path: string, params: T): string {
  return path.replace(
    /:(\w+)/g,
    (_match, token: string) =>
      (params && token in params && params[token]) || '',
  );
}

export function routify<T extends string, Q extends QueryParams = never>(
  pathname: T,
): Routified<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      if (!origin && typeof window === 'undefined') {
        throw new Error('origin is not defined and no DOM is available');
      }

      const newUrl = urlObjectAssign(
        new URL(origin || window.location.origin),
        {
          pathname: interpolatedPathname,
          ...rest,
        },
      );

      searchParams.forEach((value, key) => {
        newUrl.searchParams.set(key, value);
      });

      return newUrl;
    },
  };
}
