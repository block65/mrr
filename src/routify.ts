export { asWildcardRoute, routify } from '../lib/routify.js';

export type { Routified } from '../lib/routify.js';
