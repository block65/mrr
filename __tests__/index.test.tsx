import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Link, Route, Router, Routes, useLocation } from '../lib/index.js';
import { routify } from '../lib/routify.js';

export const LocationDisplay = () => {
  const [location] = useLocation();

  return <div data-testid="location-display">{location.pathname}</div>;
};

test('basic', async () => {
  const origin = 'https://router.example.com';

  const root = routify('/');
  const login = routify('/login');
  const userView = routify('/user/:userId');

  /* const { debug } =  */ render(
    <Router origin={origin} pathname="/">
      <LocationDisplay />
      <Routes>
        <Route path={userView.path}>
          <h1 data-testid="users">You are at the users view page</h1>
        </Route>
        <Route path={login.path}>
          <h1 data-testid="login">You are at the login page</h1>
        </Route>
        <Route path="/">
          <h1 data-testid="root">You are at the root!</h1>
          <Link url={login.build({ origin })}>login</Link>
        </Route>
        <Route>
          <h1>404</h1>
          <Link url={root.build({ origin })}>Go to root</Link>
        </Route>
      </Routes>
    </Router>,
  );

  await waitFor(() => screen.getByTestId('root'));

  fireEvent.click(screen.getByText('login'));

  await waitFor(() => screen.getByTestId('login'));

  expect(screen.getByRole('heading')).toHaveTextContent(
    'You are at the login page',
  );
});
