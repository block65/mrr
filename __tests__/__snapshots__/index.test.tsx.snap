// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
<DocumentFragment>
  <div
    data-testid="location-display"
  >
    /login
  </div>
  <h1
    data-testid="login"
  >
    You are at the login page
  </h1>
</DocumentFragment>
`;
